### DISCUSSION TITLE
Add contextmanager support to python decorator

### DISCUSSION THREAD

**Discussion by yaniv-aknin** [langfuse/langfuse#2488] [Ideas]
### Describe the feature or potential improvement

I find the low-level interface too low level, and the decorator interface too high level.

Typically decorators execute at import time and that isn't a convenient time to prepare identifiers like `session_id` and `user_id`. I can circumvent it in various ways (update the context inside the function, nested function with the inner one being decorated, etc) but that adds needless boilerplate.

So I'd recommend you add a context manager such that I can do `with langfuse.observed() as trace:` (optionally passing arguments like `name`, `as_type`, or `used_id` to `observed()`), and have the inner codeblock be reported to langfuse and manage context as usual.

It'd also be nice to have `flush=True` to force flushing when the context manager exits.

### Additional information

Thanks for langfuse!

**Comment by hassiebp** [comment]
Thanks a lot @yaniv-aknin for this push! We have added it to the ideas board for our future releases! 👍🏾

**Comment by marcklingen** [comment]
Additional notes on this by @eliorc, https://github.com/orgs/langfuse/discussions/3725

> The `observe` decorators is really useful, and on many occasions using it as a context manager makes a lot of sense.
> For example, I might have in the same function multiple LLM calls which I would like to name for better auditibility, and it would be so convinient to just 
> 
> ```python
> 
> @observe(name='my_container')
> def func(...)
>     
>       with observe(as_type='generation', name='step 1'):
>              # First LLM call
> 
>       with observe(as_type='generation', name='step 2'):
>              # Second LLM call
> ```
> 
> instead of having to put a single LLM call code line inside a function just to be able to use the `observe` decorator.
> 
> Since Python 3.2, decorators implemented using `@contextlib.contextmanager` can be used both as context managers and as decorators. I've looked at the implementation of `observe` and it is quite complex for me to try and do it on my own.
> If it's implementation could utilize `@contextlib.contextmanager` that would be awesome or alternatively implement a separate `observe_context` context manager.

**Comment by marcklingen** [comment]
Thanks again for your feedback, the latest python sdk has support for context managers: https://langfuse.com/docs/sdk/python/sdk-v3#context-managers
